consensus/safety-rules/src/network.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![allow(dead_code)]
+
+use anyhow::{anyhow, Result};
+use std::{
+    io::{Read, Write},
+    net::{SocketAddr, TcpListener, TcpStream},
+};
+
+pub struct NetworkClient {
+    stream: NetworkStream,
+}
+
+impl NetworkClient {
+    pub fn connect(server: SocketAddr) -> Result<Self> {
+        let stream = TcpStream::connect(server)?;
+        Ok(Self {
+            stream: NetworkStream::new(stream),
+        })
+    }
+
+    /// Blocking read until able to successfully read an entire message
+    pub fn read(&mut self) -> Result<Vec<u8>> {
+        self.stream.read()
+    }
+
+    /// Blocking read until able to successfully read an entire message
+    pub fn write(&mut self, data: &[u8]) -> Result<()> {
+        self.stream.write(data)
+    }
+}
+
+pub struct NetworkServer {
+    listener: TcpListener,
+    stream: Option<NetworkStream>,
+}
+
+impl NetworkServer {
+    pub fn new(listen: SocketAddr) -> Self {
+        let listener = TcpListener::bind(listen).unwrap();
+        Self {
+            listener,
+            stream: None,
+        }
+    }
+
+    /// If there isn't already a downstream client, it accepts. Otherwise it
+    /// blocks until able to successfully read an entire message
+    pub fn read(&mut self) -> Result<Vec<u8>> {
+        let stream = self.client()?;
+        stream.read()
+    }
+
+    pub fn write(&mut self, data: &[u8]) -> Result<()> {
+        let stream = self.client()?;
+        stream.write(data)
+    }
+
+    fn client(&mut self) -> Result<&mut NetworkStream> {
+        if self.stream.is_none() {
+            let (stream, _stream_addr) = self.listener.accept()?;
+            self.stream = Some(NetworkStream::new(stream));
+        }
+
+        self.stream
+            .as_mut()
+            .ok_or_else(|| anyhow!("Unable to retrieve a stream"))
+    }
+}
+
+struct NetworkStream {
+    stream: TcpStream,
+    buffer: Vec<u8>,
+    temp_buffer: [u8; 1024],
+}
+
+impl NetworkStream {
+    pub fn new(stream: TcpStream) -> Self {
+        Self {
+            stream,
+            buffer: Vec::new(),
+            temp_buffer: [0; 1024],
+        }
+    }
+
+    /// Blocking read until able to successfully read an entire message
+    pub fn read(&mut self) -> Result<Vec<u8>> {
+        let result = self.read_buffer();
+        if !result.is_empty() {
+            return Ok(result);
+        }
+
+        loop {
+            let read = self.stream.read(&mut self.temp_buffer)?;
+            self.buffer.extend(self.temp_buffer[0..read].to_vec());
+            let result = self.read_buffer();
+            if !result.is_empty() {
+                return Ok(result);
+            }
+        }
+    }
+
+    /// Blocking write until able to successfully send an entire message
+    pub fn write(&mut self, data: &[u8]) -> Result<()> {
+        let data_len = data.len() as u32;
```

Comment:
is this a safe cast?

Reply:
rewrote it to first ensure that it is within bounds and then cast.